test/OpenTelemetry.Tests/Shared/CustomDistributedContextPropagator.cs

```diff
@@ -0,0 +1,264 @@
+
+                if (baggageString[currentIndex] != '=')
+                {
+                    // Skip Spaces
+                    while (currentIndex < baggageString.Length && (baggageString[currentIndex] == Space || baggageString[currentIndex] == Tab))
+                    {
+                        currentIndex++;
+                    }
+
+                    if (currentIndex >= baggageString.Length)
+                    {
+                        break; // Wrong key format
+                    }
+
+                    if (baggageString[currentIndex] != '=')
+                    {
+                        break; // wrong key format.
+                    }
+                }
+
+                currentIndex++;
+
+                // Skip spaces
+                while (currentIndex < baggageString.Length && (baggageString[currentIndex] == Space || baggageString[currentIndex] == Tab))
+                {
+                    currentIndex++;
+                }
+
+                if (currentIndex >= baggageString.Length)
+                {
+                    break; // Wrong value format
+                }
+
+                int valueStart = currentIndex;
+
+                // Search end of the value
+                while (currentIndex < baggageString.Length && baggageString[currentIndex] != Space && baggageString[currentIndex] != Tab &&
+                       baggageString[currentIndex] != Comma && baggageString[currentIndex] != Semicolon)
+                {
+                    currentIndex++;
+                }
+
+                if (keyStart < keyEnd && valueStart < currentIndex)
+                {
+                    baggageList ??= new List<KeyValuePair<string, string?>>();
+
+                    // Insert in reverse order for asp.net compatibility.
+                    baggageList.Insert(0, new KeyValuePair<string, string?>(
+                                                WebUtility.UrlDecode(baggageString.Substring(keyStart, keyEnd - keyStart)).Trim(STrimmingSpaceCharacters),
+                                                WebUtility.UrlDecode(baggageString.Substring(valueStart, currentIndex - valueStart)).Trim(STrimmingSpaceCharacters)));
+                }
+
+                // Skip to end of values
+                while (currentIndex < baggageString.Length && baggageString[currentIndex] != Comma)
+                {
+                    currentIndex++;
+                }
+
+                currentIndex++; // Move to next key-value entry
+            }
+            while (currentIndex < baggageString.Length);
+
+            baggage = baggageList;
+            return baggageList != null;
+        }
+
+        private void InjectBaggage(Baggage currentBaggage, object? carrier, PropagatorSetterCallback setter)
+        {
+            using var e = currentBaggage.GetEnumerator();
+
+            if (e.MoveNext() == true)
+            {
+                int itemCount = 0;
+                StringBuilder baggage = new StringBuilder();
```

Comment:
If you want some ideas for how you could improve the perf here, check out: https://github.com/dotnet/runtime/blob/fd31f1f01a1c755c965e6f20b3dfd08933e27808/src/libraries/Common/src/System/Text/ValueStringBuilder.cs#L11